haskell-api.md

```diff
@@ -0,0 +1,217 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+- If filtering on an attribute of the main resource: `{attribute}={value}`
+- If filtering on an attribute of a related resource:
+  `{resource}.{attribute}={value}`
+- If one thing has many of another, pluralize the many
+
+Examples, for a `/teachers` endpoint:
+
+- `?name=`, because a Teacher has one name
+- `?school.id=`, because School is a relation, and a Teacher has only one
```

Comment:
I think it is easier to consider relations always plural. A teachers endpoint may pull back many teachers and they may have many `schools`. Relations are always plural is much simpler than anything else.

Reply:
I did some more thinking on this because I agree it's got a downside in complexity as written.

However, I think it's actually simpler to just say it should match a path in the response:

```
/teachers
[ { id:
  , school: { id: }
  }
]
```

Therefore, `/teachers?school.id=`.

```
/students
[ { id:
  , schools: [{ id:, ... }]
  }
]
```

Therefore, `/students?schools.id=`.

This is the same exact outcome, but a lot simpler to state as a guide. The only gap is when you are filtering by something not in the response, which we will likely support, but can (IMO) be inferred by how it *would* look in a response.

WDYT?

Reply:
Agreed, that is a clearer and more intuitive criteria since it matches the resource and the filter.